<a href="https://colab.research.google.com/github/yeticheese/NMA_2024_Zealous_Sedum_CN/blob/main/notebooks/A_Preprocessing.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
# @markdown Install packages
!pip install pip --upgrade --quiet
!pip install brain_observatory_utilities --upgrade --quiet
!pip install pandas --quiet
!pip install seaborn --quiet

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 12.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.9/58.9 kB 2.6 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 49.3/49.3 kB 3.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.1/3.1 MB 61.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.1/12.1 MB 86.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 57.6/57.6 kB 3.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 336.0/336.0 kB 19.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 17.1/17.1 MB 85.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.7/62.7 kB 4.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 34.4/34.4 MB 15.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 973.1/973.1 kB 38.4 MB/s eta 0:00

In [3]:
#@markdown Import Libraries
import os
import numpy as np
import pandas as pd
from tqdm import tqdm
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn import svm
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix, accuracy_score
from sklearn.manifold import TSNE

import brain_observatory_utilities.datasets.optical_physiology.data_formatting as ophys_formatting
import brain_observatory_utilities.utilities.general_utilities as utilities

from allensdk.brain_observatory.behavior.behavior_project_cache import VisualBehaviorOphysProjectCache

pd.set_option('display.max_columns', 500)
# this line may be needed if you run into Error in pandas query function
# Otherwise set the engine to python in queries made throught the book
# pd.DataFrame.query = lambda self, expr, **kwargs: self.query(expr, engine='python', **kwargs)

# Abstract

// To be filled


In [4]:
#@title Data Retrival
# @title Data retrieval
import os, requests

fname = "allen_visual_behavior_2p_change_detection_familiar_novel_image_sets.parquet"
url = "https://ndownloader.figshare.com/files/28470255"

if not os.path.isfile(fname):
  try:
    r = requests.get(url)
  except requests.ConnectionError:
    print("!!! Failed to download data !!!")
  else:
    if r.status_code != requests.codes.ok:
      print("!!! Failed to download data !!!")
    else:
      with open(fname, "wb") as fid:
        fid.write(r.content)

In [5]:
filename = "allen_visual_behavior_2p_change_detection_familiar_novel_image_sets.parquet"
data = pd.read_parquet(filename)

### Displaying the data head

In [11]:
data.head(10)

stimulus_presentations_id  cell_specimen_id  \
85                          5        1086496689   
86                          5        1086501573   
87                          5        1086500533   
88                          5        1086497438   
89                          5        1086497908   
90                          5        1086501720   
91                          5        1086497129   
92                          5        1086497034   
93                          5        1086496500   
94                          5        1086498071   

                                                trace  \
85  [0.06366926431655884, 0.002476318972185254, -0...   
86  [0.0637083426117897, 0.0919334813952446, -0.09...   
87  [0.0223990511149168, 0.22551870346069336, 0.11...   
88  [0.01795899122953415, 0.15271763503551483, 0.0...   
89  [0.059766266494989395, 0.03602207452058792, 0....   
90  [0.1370878517627716, 0.06181945651769638, 0.14...   
91  [-0.01640218496322632, -0.05918422341346741, -...   
92  [0.12177210301160812, 0.07332494109869003, -0....   
93  [-0.1013624370098114, -0.06485532969236374, 0....   
94  [0.0267901923507452, -0.04602086544036865, 0.0...   

                                     trace_timestamps  mean_response  \
85  [-1.228133913303992, -1.1958145997959921, -1.1...       0.006115   
86  [-1.228133913303992, -1.1958145997959921, -1.1...      -0.018832   
87  [-1.228133913303992, -1.1958145997959921, -1.1...      -0.041275   
88  [-1.228133913303992, -1.1958145997959921, -1.1...       0.039527   
89  [-1.228133913303992, -1.1958145997959921, -1.1...      -0.006628   
90  [-1.228133913303992, -1.1958145997959921, -1.1...       0.010061   
91  [-1.228133913303992, -1.1958145997959921, -1.1...       0.087167   
92  [-1.228133913303992, -1.1958145997959921, -1.1...       0.001650   
93  [-1.228133913303992, -1.1958145997959921, -1.1...      -0.003985   
94  [-1.228133913303992, -1.1958145997959921, -1.1...       0.016455   

    baseline_response image_name  image_index  is_change  omitted  \
85          -0.084113      im075            2       True    False   
86          -0.025578      im075            2       True    False   
87          -0.032698      im075            2       True    False   
88          -0.026750      im075            2       True    False   
89          -0.039277      im075            2       True    False   
90          -0.113912      im075            2       True    False   
91          -0.084272      im075            2       True    False   
92           0.019924      im075            2       True    False   
93          -0.011439      im075            2       True    False   
94          -0.013882      im075            2       True    False   

    mean_running_speed  mean_pupil_area  response_latency  rewarded  \
85            0.000006      5822.220145            0.5004      True   
86            0.000006      5822.220145            0.5004      True   
87            0.000006      5822.220145            0.5004      True   
88            0.000006      5822.220145            0.5004      True   
89            0.000006      5822.220145            0.5004      True   
90            0.000006      5822.220145            0.5004      True   
91            0.000006      5822.220145            0.5004      True   
92            0.000006      5822.220145            0.5004      True   
93            0.000006      5822.220145            0.5004      True   
94            0.000006      5822.220145            0.5004      True   

    ophys_experiment_id  imaging_depth targeted_structure      cre_line  \
85            994278281            275               VISp  Sst-IRES-Cre   
86            994278281            275               VISp  Sst-IRES-Cre   
87            994278281            275               VISp  Sst-IRES-Cre   
88            994278281            275               VISp  Sst-IRES-Cre   
89            994278281            275               VISp  Sst-IRES-Cre   
90            994278281      

In [39]:
#@markdown Let's find out if we should consider targeted areas
print(f'{data.targeted_structure.unique()[0]} is the only area in our dataset')

VISp is the only area in our dataset


## Dimensions to consider
- `cell_specimen_id`
- `trace`
- `trace_timestamps`
- `mean_response`
- `image_name`
- `image_index` (maybe)
- `omitted`
- `mean_running_speed`
- `mean_pupil_area`
- `rewarded`
- `imaging_depth`
- `cre_line`
- `exposure_level`

## Housekeeping labels to consider
-`mouse_id`

In [32]:
trimmed_data = data[['stimulus_presentations_id','cell_specimen_id','trace','trace_timestamps','mean_response','image_index','omitted','mean_running_speed','mean_pupil_area','rewarded','imaging_depth','cre_line','exposure_level','mouse_id']]
trimmed_data.head(10)

stimulus_presentations_id  cell_specimen_id  \
85                          5        1086496689   
86                          5        1086501573   
87                          5        1086500533   
88                          5        1086497438   
89                          5        1086497908   
90                          5        1086501720   
91                          5        1086497129   
92                          5        1086497034   
93                          5        1086496500   
94                          5        1086498071   

                                                trace  \
85  [0.06366926431655884, 0.002476318972185254, -0...   
86  [0.0637083426117897, 0.0919334813952446, -0.09...   
87  [0.0223990511149168, 0.22551870346069336, 0.11...   
88  [0.01795899122953415, 0.15271763503551483, 0.0...   
89  [0.059766266494989395, 0.03602207452058792, 0....   
90  [0.1370878517627716, 0.06181945651769638, 0.14...   
91  [-0.01640218496322632, -0.05918422341346741, -...   
92  [0.12177210301160812, 0.07332494109869003, -0....   
93  [-0.1013624370098114, -0.06485532969236374, 0....   
94  [0.0267901923507452, -0.04602086544036865, 0.0...   

                                     trace_timestamps  mean_response  \
85  [-1.228133913303992, -1.1958145997959921, -1.1...       0.006115   
86  [-1.228133913303992, -1.1958145997959921, -1.1...      -0.018832   
87  [-1.228133913303992, -1.1958145997959921, -1.1...      -0.041275   
88  [-1.228133913303992, -1.1958145997959921, -1.1...       0.039527   
89  [-1.228133913303992, -1.1958145997959921, -1.1...      -0.006628   
90  [-1.228133913303992, -1.1958145997959921, -1.1...       0.010061   
91  [-1.228133913303992, -1.1958145997959921, -1.1...       0.087167   
92  [-1.228133913303992, -1.1958145997959921, -1.1...       0.001650   
93  [-1.228133913303992, -1.1958145997959921, -1.1...      -0.003985   
94  [-1.228133913303992, -1.1958145997959921, -1.1...       0.016455   

    image_index  omitted  mean_running_speed  mean_pupil_area  rewarded  \
85            2    False            0.000006      5822.220145      True   
86            2    False            0.000006      5822.220145      True   
87            2    False            0.000006      5822.220145      True   
88            2    False            0.000006      5822.220145      True   
89            2    False            0.000006      5822.220145      True   
90            2    False            0.000006      5822.220145      True   
91            2    False            0.000006      5822.220145      True   
92            2    False            0.000006      5822.220145      True   
93            2    False            0.000006      5822.220145      True   
94            2    False            0.000006      5822.220145      True   

    imaging_depth      cre_line exposure_level mouse_id  
85            275  Sst-IRES-Cre       familiar   476970  
86            275  Sst-IRES-Cre       familiar   476970  
87            275  Sst-IRES-Cre       familiar   476970  
88            275  Sst-IRES-Cre       familiar   476970  
89            275  Sst-IRES-Cre       familiar   476970  
90            275  Sst-IRES-Cre       familiar   476970  
91            275  Sst-IRES-Cre       familiar   476970  
92            275  Sst-IRES-Cre       familiar   476970  
93            275  Sst-IRES-Cre       familiar   476970  
94            275  Sst-IRES-Cre       familiar   476970

In [43]:
trimmed_data

stimulus_presentations_id  cell_specimen_id  \
85                               5        1086496689   
86                               5        1086501573   
87                               5        1086500533   
88                               5        1086497438   
89                               5        1086497908   
...                            ...               ...   
1709437                       4795        1086498401   
1709438                       4795        1086498544   
1709439                       4795        1086498699   
1709440                       4795        1086498889   
1709441                       4795        1086498976   

                                                     trace  \
85       [0.06366926431655884, 0.002476318972185254, -0...   
86       [0.0637083426117897, 0.0919334813952446, -0.09...   
87       [0.0223990511149168, 0.22551870346069336, 0.11...   
88       [0.01795899122953415, 0.15271763503551483, 0.0...   
89       [0.059766266494989395, 0.03602207452058792, 0....   
...                                                    ...   
1709437  [0.04858933016657829, -0.05913962423801422, 0....   
1709438  [-0.02712165005505085, 0.0, 0.0054037026129662...   
1709439  [0.04359331354498863, -0.010188717395067215, 0...   
1709440  [-0.0878591388463974, -0.013658315874636173, 0...   
1709441  [0.01386266853660345, -0.017286673188209534, 0...   

                                          trace_timestamps  mean_response  \
85       [-1.228133913303992, -1.1958145997959921, -1.1...       0.006115   
86       [-1.228133913303992, -1.1958145997959921, -1.1...      -0.018832   
87       [-1.228133913303992, -1.1958145997959921, -1.1...      -0.041275   
88       [-1.228133913303992, -1.1958145997959921, -1.1...       0.039527   
89       [-1.228133913303992, -1.1958145997959921, -1.1...      -0.006628   
...                                                    ...            ...   
1709437  [-1.2279264819932727, -1.1956126272039762, -1....       0.021412   
1709438  [-1.2279264819932727, -1.1956126272039762, -1....      -0.006323   
1709439  [-1.2279264819932727, -1.1956126272039762, -1....      -0.036777   
1709440  [-1.2279264819932727, -1.1956126272039762, -1....      -0.042176   
1709441  [-1.2279264819932727, -1.1956126272039762, -1....       0.003847   

         image_index  omitted  mean_running_speed  mean_pupil_area  rewarded  \
85                 2    False            0.000006      5822.220145      True   
86                 2    False            0.000006      5822.220145      True   
87                 2    False            0.000006      5822.220145      True   
88                 2    False            0.000006      5822.220145      True   
89                 2    False            0.000006      5822.220145      True   
...              ...      ...                 ...              ...       ...   
1709437            8     True           12.433663     13239.100342     False   
1709438            8     True           12.433663     13239.100342     False   
1709439            8     True           12.433663     13239.100342     False   
1709440            8     True           12.433663     13239.100342     False   
1709441            8     True           12.433663     13239.100342     False   

         imaging_depth      cre_line exposure_level mouse_id  
85                 275  Sst-IRES-Cre       familiar   476970  
86                 275  Sst-IRES-Cre       familiar   476970  
87                 275  Sst-IRES-Cre       familiar   476970  
88                 275  Sst-IRES-Cre       familiar   476970  
89                 275  Sst-IRES-Cre       familiar   476970  
...                ...           ...            ...      ...  
1709437            175  Vip-IRES-Cre          novel   467954  
1709438            175  Vip-IRES-Cre          novel   467954  
1709439            175  Vip-IRES-Cre          novel   467954  
1709440            175  Vip-IRES-Cre          novel   467954  
1709441     

In [33]:
len(trimmed_data.cell_specimen_id.unique())

223

In [34]:
len(trimmed_data.index.unique())

147695

## Resulting dataset
The Dataset should be at least 223 columns, each column for each `cell_specimen_id` and 147695 rows for each `stimulus_presentations_id`

In [40]:
features_labels = trimmed_data.pivot(columns='cell_specimen_id',values='mean_response')

In [41]:
features_labels.head(20)

cell_specimen_id  1086489847  1086489860  1086489882  1086489891  1086489901  \
85                       NaN         NaN         NaN         NaN         NaN   
86                       NaN         NaN         NaN         NaN         NaN   
87                       NaN         NaN         NaN         NaN         NaN   
88                       NaN         NaN         NaN         NaN         NaN   
89                       NaN         NaN         NaN         NaN         NaN   
90                       NaN         NaN         NaN         NaN         NaN   
91                       NaN         NaN         NaN         NaN         NaN   
92                       NaN         NaN         NaN         NaN         NaN   
93                       NaN         NaN         NaN         NaN         NaN   
94                       NaN         NaN         NaN         NaN         NaN   
95                       NaN         NaN         NaN         NaN         NaN   
96                       NaN         NaN         NaN         NaN         NaN   
97                       NaN         NaN         NaN         NaN         NaN   
98                       NaN         NaN         NaN         NaN         NaN   
99                       NaN         NaN         NaN         NaN         NaN   
100                      NaN         NaN         NaN         NaN         NaN   
101                      NaN         NaN         NaN         NaN         NaN   
187                      NaN         NaN         NaN         NaN         NaN   
188                      NaN         NaN         NaN         NaN         NaN   
189                      NaN         NaN         NaN         NaN         NaN   

cell_specimen_id  1086490391  1086490482  1086490519  1086490526  1086490561  \
85                       NaN         NaN         NaN         NaN         NaN   
86                       NaN         NaN         NaN         NaN         NaN   
87                       NaN         NaN         NaN         NaN         NaN   
88                       NaN         NaN         NaN         NaN         NaN   
89                       NaN         NaN         NaN         NaN         NaN   
90                       NaN         NaN         NaN         NaN         NaN   
91                       NaN         NaN         NaN         NaN         NaN   
92                       NaN         NaN         NaN         NaN         NaN   
93                       NaN         NaN         NaN         NaN         NaN   
94                       NaN         NaN         NaN         NaN         NaN   
95                       NaN         NaN         NaN         NaN         NaN   
96                       NaN         NaN         NaN         NaN         NaN   
97                       NaN         NaN         NaN         NaN         NaN   
98                       NaN         NaN         NaN         NaN         NaN   
99                       NaN         NaN         NaN         NaN         NaN   
100                      NaN         NaN         NaN         NaN         NaN   
101                      NaN         NaN         NaN         NaN         NaN   
187                      NaN         NaN         NaN         NaN         NaN   
188                      NaN         NaN         NaN         NaN         NaN   
189                      NaN         NaN         NaN         NaN         NaN   

cell_specimen_id  1086490572  1086490598  1086490603  1086490613  1086490648  \
85                       NaN         NaN         NaN         NaN         NaN   
86                       NaN         NaN         NaN         NaN         NaN   
87                       NaN         NaN         NaN         NaN         NaN   
88                       NaN         NaN         NaN         NaN         NaN   
89                       NaN         NaN         NaN         NaN         NaN   
90                       NaN         NaN         NaN         NaN         NaN   
91                       NaN         NaN         NaN         NaN         NaN  

# Dimensionality Reduction

In [ ]:
# Farzad's PCA will go here